provider_azure.go

```diff
@@ -0,0 +1,185 @@
+// In the future this could be used with other providers as well. Example:
+//
+// {
+// 	 "_claim_names": {
+// 	   "groups": "src1"
+// 	 },
+// 	 "_claim_sources": {
+// 	   "src1": {
+// 	     "endpoint": "https://graph.windows.net...."
+// 	   }
+//   }
+// }
+//
+// For this to work, "profile" should be set in "oidc_scopes" in the vault oidc role.
+//
+func (a *AzureProvider) getClaimSource(logger log.Logger, allClaims map[string]interface{}, role *jwtRole) (string, error) {
+	// Get the source key for the groups claim
+	name := fmt.Sprintf("/%s/%s", claimNamesField, role.GroupsClaim)
+	groupsClaimSource := getClaim(logger, allClaims, name)
+	if groupsClaimSource == nil {
+		return "", fmt.Errorf("unable to locate groups claim %q in %s", role.GroupsClaim, claimNamesField)
+	}
+	// Get the endpoint source for the groups claim
+	endpoint := fmt.Sprintf("/%s/%s/endpoint", claimSourcesField, groupsClaimSource.(string))
+	val := getClaim(logger, allClaims, endpoint)
+	if val == nil {
+		return "", fmt.Errorf("unable to locate %s in claims", endpoint)
+	}
+	logger.Debug(fmt.Sprintf("found Azure Graph API endpoint for group membership: %v", val))
+	return fmt.Sprintf("%v", val), nil
+}
+
+// Fetch user groups from the Azure AD Graph API
+func (a *AzureProvider) getAzureGroups(logger log.Logger, groupsURL string, c *jwtConfig) (interface{}, error) {
+	urlParsed, err := url.Parse(groupsURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse distributed groups source url %s: %s", groupsURL, err)
+	}
+	token, err := a.getAzureToken(logger, c, urlParsed.Host)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get token: %s", err)
+	}
+	payload := strings.NewReader("{\"securityEnabledOnly\": false}")
+	req, _ := http.NewRequest("POST", groupsURL, payload)
+	req.Header.Add("content-type", "application/json")
+	req.Header.Add("authorization", fmt.Sprintf("Bearer %s", token))
+
+	// If endpoint is the old windows graph api, add api-version
+	if urlParsed.Host == windowsGraphHost {
+		query := req.URL.Query()
+		query.Add("api-version", windowsAPIVersion)
+		req.URL.RawQuery = query.Encode()
+	}
+	client := http.DefaultClient
+	if c, ok := a.ctx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		client = c
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("unable to call Azure AD Graph API: %s", err)
+	}
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read Azure AD Graph API response: %s", err)
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to get groups: %s", string(body))
+	}
+
+	var target azureGroups
+	if err := json.Unmarshal(body, &target); err != nil {
+		return nil, fmt.Errorf("unabled to decode response: %s", err)
+	}
+	return target.Value, nil
+}
+
+// Login to Azure, using client id and secret.
+func (a *AzureProvider) getAzureToken(logger log.Logger, c *jwtConfig, host string) (string, error) {
```

Comment:
nit: logger is unused here, so you could remove it.

Reply:
Removed in 777e0ae